File: IK_numerico.py
import numpy as np
from scipy import io, integrate, linalg, signal
import Articulado

def movement(init,target):
  N = 10
  Caminho = np.array(
      [-1,# * np.ones(N),
        0,# * np.ones(N),
        0,# * np.ones(N),
        
        0,# * np.ones(N),
        1,# * np.ones(N),
        0,# * np.ones(N),
              
        0,# * np.ones(N),
        0,# * np.ones(N),
        -1,# * np.ones(N),
              
        target[0],#np.linspace(init[0], target[0], num=N),
        target[1],#np.linspace(init[1], target[1], num=N),
        target[2]])#np.linspace(init[2], target[2], num=N)])
  Caminho = Caminho.reshape(12,1)

  L1 = 0.0661+0.0085
  L2 = 0.4251
  L3 = 0.3922
  L4 = 0.0397+0.0703
  L5 = 0.0492+0.04545
  L6 = 0.0996-0.02465

  th1 = np.deg2rad(90)
  th2 = np.deg2rad(-60)
  th3 = np.deg2rad(100)
  th4 = np.deg2rad(140)
  th5 = np.deg2rad(180)
  th6 = np.deg2rad(90)
  '''th1 = np.deg2rad(0)
  th2 = np.deg2rad(0)
  th3 = np.deg2rad(0)
  th4 = np.deg2rad(0)
  th5 = np.deg2rad(0)
  th6 = np.deg2rad(0)'''
  Th = np.array([th1,th2,th3,th4,th5,th6])
  Th = Th.reshape(6,1)
  TH=Th
  F0 = np.eye(4)
  F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)

  for index,x in np.ndenumerate(F6):
    if index[0]==3:
      break
    elif index==(0,0):
      fe=np.array([x])
    else:
      fe=np.append(fe,np.array([x]),axis=0)

  fe = fe.reshape(12,1)
  n = 0.2
  ksi= (0.01)**2

  for k in range(Caminho.shape[1]):
    fg = Caminho.T[k]
    fg = fg.reshape(12,1)
    F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    aux = np.append(F6[0:3,0],np.append(F6[0:3,1],np.append(F6[0:3,2],F6[0:3,3])))
    aux = aux.reshape(12,1)
    fe = np.concatenate((fe,aux),1)
    aux = fe[:,-1]
    aux = aux.reshape(12,1)
    E = 0.5 * (fg - aux).T @ (fg - aux)
    while E.item() > ksi :
      dE = -(fg-aux)
      J = Articulado.ArticuladoJacobiano(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      dTh = linalg.pinv(J) @ dE
      Th = Th - n * dTh
      th1 = Th.item(0)
      th2 = Th.item(1)
      th3 = Th.item(2)
      th4 = Th.item(3)
      th5 = Th.item(4)
      th6 = Th.item(5)
      F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
      aux = np.append(F6[0:3,0],np.append(F6[0:3,1],np.append(F6[0:3,2],F6[0:3,3])))
      aux = aux.reshape(12,1)
      fe = np.concatenate((fe,aux),1)
      aux = fe[:,-1]
      aux = aux.reshape(12,1)
      E = 0.5 * (fg - aux).T @ (fg - aux)
      print("Error:",E,"- ksi:",k)
    TH = np.concatenate((TH,Th),1)
  return(TH.transpose()[-1])


'''
N = np.arange(-1,1,0.1)
Caminho1 = np.array(
    [-1 * np.ones(N.size),
      0 * np.ones(N.size),
      0 * np.ones(N.size),
       
      0 * np.ones(N.size),
      1 * np.ones(N.size),
      0 * np.ones(N.size),
            
      0 * np.ones(N.size),
      0 * np.ones(N.size),
     -1 * np.ones(N.size),
            
     -0.5 * np.ones(N.size),
     -0.350 * N,
      0.25 * np.ones(N.size)])

Caminho2 = np.array(
    [-1 * np.ones(N.size),
      0 * np.ones(N.size),
      0 * np.ones(N.size),
       
      0 * np.ones(N.size),
      1 * np.ones(N.size),
      0 * np.ones(N.size),
            
      0 * np.ones(N.size),
      0 * np.ones(N.size),
     -1 * np.ones(N.size),
            
      0.5 * np.ones(N.size),
     -0.350 * N,
      0.25 * np.ones(N.size)])

Caminho = np.append(Caminho1,Caminho2,axis=1)
L1 = 0.0661+0.0085
L2 = 0.4251
L3 = 0.3922
L4 = 0.0397+0.0703
L5 = 0.0492+0.04545
L6 = 0.0996-0.02465

th1 = np.deg2rad(90)
th2 = np.deg2rad(-60)
th3 = np.deg2rad(100)
th4 = np.deg2rad(140)
th5 = np.deg2rad(180)
th6 = np.deg2rad(90)
Th = np.array([th1,th2,th3,th4,th5,th6])
Th = Th.reshape(6,1)
TH=Th
F0 = np.eye(4)
F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)

for index,x in np.ndenumerate(F6):
  if index[0]==3:
    break
  elif index==(0,0):
    fe=np.array([x])
  else:
    fe=np.append(fe,np.array([x]),axis=0)

fe = fe.reshape(12,1)
n = 0.2
ksi= (0.01)**2

for k in range(Caminho.shape[1]):
  fg = Caminho.T[k]
  fg = fg.reshape(12,1)
  F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
  aux = np.append(F6[0:3,0],np.append(F6[0:3,1],np.append(F6[0:3,2],F6[0:3,3])))
  aux = aux.reshape(12,1)
  fe = np.concatenate((fe,aux),1)
  aux = fe[:,-1]
  aux = aux.reshape(12,1)
  E = 0.5 * (fg - aux).T @ (fg - aux)
  while E.item() > ksi :
    dE = -(fg-aux)
    J = Articulado.ArticuladoJacobiano(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    dTh = linalg.pinv(J) @ dE
    Th = Th - n * dTh
    th1 = Th.item(0)
    th2 = Th.item(1)
    th3 = Th.item(2)
    th4 = Th.item(3)
    th5 = Th.item(4)
    th6 = Th.item(5)
    F1 = Articulado.ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F2 = Articulado.ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F3 = Articulado.ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F4 = Articulado.ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F5 = Articulado.ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    F6 = Articulado.ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6)
    aux = np.append(F6[0:3,0],np.append(F6[0:3,1],np.append(F6[0:3,2],F6[0:3,3])))
    aux = aux.reshape(12,1)
    fe = np.concatenate((fe,aux),1)
    aux = fe[:,-1]
    aux = aux.reshape(12,1)
    E = 0.5 * (fg - aux).T @ (fg - aux)
  TH = np.concatenate((TH,Th),1)
print(TH)
'''

File: matlab/IK_numerico.m
% ==============================
% Cinemática Inversa do robô RRR
% ==============================
clear all; clc; close all;
% Caminho a ser percorrido pelo robô
N = -1:0.1:1;
Caminho1 = [-1* ones(size(N));
            0* ones(size(N));
            0* ones(size(N));
            
            0* ones(size(N));
            1* ones(size(N));
            0* ones(size(N));
            
            0* ones(size(N));
            0* ones(size(N));
           -1* ones(size(N));
            
           -0.5 * ones(size(N));
           -0.350 * N;
           0.25 * ones(size(N))];

Caminho2 = [-1* ones(size(N));
            0* ones(size(N));
            0* ones(size(N));
            
            0* ones(size(N));
            1* ones(size(N));
            0* ones(size(N));
            
            0* ones(size(N));
            0* ones(size(N));
           -1* ones(size(N));
            
           0.5 * ones(size(N));
           -0.350 * N;
           0.25 * ones(size(N))];
Caminho = [Caminho1,Caminho2];
% Comprimento dos links
L1 = 0.0661+0.0085; L2 = 0.4251; L3 = 0.3922; L4 = 0.0397+0.0703; L5 = 0.0492+0.04545; L6 = 0.0996-0.02465;
% Inicializando o robô
th1 = deg2rad(90); th2 = deg2rad(-60); th3 = deg2rad(100); th4 = deg2rad(140); th5 = deg2rad(180); th6 = deg2rad(90); Th = [th1;th2;th3;th4;th5;th6]; TH=Th;
F0 = eye(4);
F1 = ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
F2 = ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
F3 = ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
F4 = ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
F5 = ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
F6 = ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);

% Função da posição do end-effectordo robô
fe = [F6(1:3,1);F6(1:3,2);F6(1:3,3);F6(1:3,4)];
% Taxa de aprendizado
n = 0.2;
% Tolerância
ksi= (0.01)^2;

% == LOOP DE OTIMIZACAO ==
% Para cada um dos pontos do caminho
for k = 1:size(Caminho,2)
    % Obtém o objetivo atual
    fg = Caminho(:,k);
    % Calcula a posição atual do robô
    F1 = ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    F2 = ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    F3 = ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    F4 = ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    F5 = ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    F6 = ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
    % Função da posição do end-effectordo robô
    fe = [fe , [F6(1:3,1);F6(1:3,2);F6(1:3,3);F6(1:3,4)]];
    % Erro quadrático médio inicial
    E = 1/2 * (fg -fe(:,end))' * (fg -fe(:,end));
    % Loop para convergir ao ponto desejado
    while (E > ksi)
        % Calcula o Gradiente cartesiano atual
        dE = -(fg-fe(:,end));
        % Calcula o Jacobiano para a posicaoangular corrente
        J = ArticuladoJacobiano(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5 , L6);
        % Obtenha o Gradiente no espaço de juntas observando a degeneração
        dTh = pinv(J) * dE;
        % Atualize o valor das juntas
        Th = Th - n * dTh;
        % Atualiza o robô
        th1 = Th(1); th2 = Th(2); th3 = Th(3); th4 = Th(4); th5 = Th(5); th6 = Th(6);
        F1 = ArticuladoF1(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        F2 = ArticuladoF2(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        F3 = ArticuladoF3(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        F4 = ArticuladoF4(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        F5 = ArticuladoF5(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        F6 = ArticuladoF6(th1 , th2 , th3, th4 , th5 , th6 , L1 , L2 , L3 , L4 , L5, L6);
        
        % Função da posição do end-effector do robô
        fe = [fe, [F6(1:3,1);F6(1:3,2);F6(1:3,3);F6(1:3,4)]];
        
        % Erro quadrático médio inicial
        E = 1/2 * (fg-fe(:,end))' * (fg-fe(:,end));
        
        % == PLOT ==
        % Caminho a ser percorrido
        plot3(Caminho(10,:) , Caminho(11,:) , Caminho(12,:) , 'ob', 'linewidth', 2 , 'markersize', 10), hold on,
        
       % Ponto objetivo atual
        plot3(Caminho(10,k) , Caminho(11,k) , Caminho(12,k) , 'or', 'linewidth', 2 , 'markersize', 10)
        
        % Caminho executado pelo robo
        plot3(fe(10,:) , fe(11,:) , fe(12,:) , 'm', 'linewidth', 2)
        
        % Robô
        PlotTransicaoAB(F0 , F1), PlotFrameA(F0), PlotFrameA(F1),
        PlotTransicaoAB(F1 , F2),PlotFrameA(F2),
        PlotTransicaoAB(F2 , F3),PlotFrameA(F3),
        PlotTransicaoAB(F3 , F4),PlotFrameA(F4),
        PlotTransicaoAB(F4 , F5),PlotFrameA(F5),
        PlotTransicaoAB(F5 , F6),PlotFrameA(F6),
        hold off, xlabel('x'); ylabel('y'); zlabel('z');
        axis equal; grid on;
        drawnow;
    end
    TH = [TH,Th];
end